sdk/storage/Azure.Storage.DataMovement/src/Shared/DataTransfer.cs

```diff
@@ -62,56 +63,29 @@ internal DataTransfer(string id, long bytesTransferred)
             _state = new DataTransferState(id, bytesTransferred);
         }
 
-        /// <summary>
-        /// Disposes the DataTransfer object.
-        /// </summary>
-        /// <returns></returns>
-        public async ValueTask DisposeAsync()
-        {
-            await _state.DisposeAsync().ConfigureAwait(false);
-            GC.SuppressFinalize(this);
-        }
-
         /// <summary>
         /// Ensures completion of the DataTransfer and attempts to get result
         /// </summary>
         public void EnsureCompleted(CancellationToken cancellationToken = default)
         {
-#if DEBUG
-            VerifyTaskCompleted(HasCompleted);
-#endif
 #pragma warning disable AZC0102 // Do not use GetAwaiter().GetResult(). Use the TaskExtensions.EnsureCompleted() extension method instead.
-            AwaitCompletion(cancellationToken);
+            AwaitCompletion(cancellationToken).GetAwaiter().GetResult();
 #pragma warning restore AZC0102 // Do not use GetAwaiter().GetResult(). Use the TaskExtensions.EnsureCompleted() extension method instead.
         }
 
-        [Conditional("DEBUG")]
-        private static void VerifyTaskCompleted(bool isCompleted)
-        {
-            if (!isCompleted)
-            {
-                if (Debugger.IsAttached)
-                {
-                    Debugger.Break();
-                }
-            }
-        }
-
         /// <summary>
         /// Waits until the data transfer itself has completed
         /// </summary>
         /// <param name="cancellationToken"></param>
-        public Task AwaitCompletion(CancellationToken cancellationToken = default)
+        public async Task AwaitCompletion(CancellationToken cancellationToken = default)
         {
-            while (!HasCompleted)
+            if ( _state.HasCompleted ||
+                (cancellationToken != default && cancellationToken.IsCancellationRequested))
             {
-#if DEBUG
-                VerifyTaskCompleted(HasCompleted);
-#endif
-                CancellationHelper.ThrowIfCancellationRequested(cancellationToken);
+                return;
             }
-
```

Comment:
You shouldn't need to special case the task being completed or cancelled. You should be able to always await the completion sources task as below.